mava/components/tf/losses/sequence.py

```diff
@@ -27,45 +26,64 @@ def recurrent_n_step_critic_loss(
     q_values: tf.Tensor,
     target_q_values: tf.Tensor,
     rewards: tf.Tensor,
-    discounts: tf.Tensor,
     bootstrap_n: int,
-    loss_fn: Union[Type[trfl.td_learning], Type[losses.categorical]],
+    discount: float,
+    end_of_episode: tf.Tensor,
 ) -> tf.Tensor:
 
+    if type(q_values) == DiscreteValuedDistribution:
+        # Convert to tensor values
+        q_values = tf.reshape(q_values._mean(), rewards.shape)
+        target_q_values = tf.reshape(target_q_values._mean(), rewards.shape)
+
+    q_values = tf.squeeze(q_values)
+
+    target_q_values = tf.squeeze(target_q_values)
+    rewards = tf.squeeze(rewards)
+    check_rank([q_values, target_q_values, rewards], [2, 2, 2])
+
     seq_len = len(rewards[0])
-    assert bootstrap_n < seq_len
-    if type(q_values) != DiscreteValuedDistribution:
-        check_rank([q_values, target_q_values, rewards, discounts], [2, 2, 2, 2])
+
+    # d: discount * done
+    # bootstrap_n=1 is the normal return of Q_t-1 = R_t-1 + d * Q_t
+    # bootstrap_n=seq_len is the Q_t_1 = discounted sum of rewards return
+    assert 0 < bootstrap_n <= len(rewards[0])
 
     # Construct arguments to compute bootstrap target.
-    if type(q_values) != DiscreteValuedDistribution:
-        q_tm1 = q_values[:, 0:-bootstrap_n]
-        q_t = target_q_values[:, bootstrap_n:]
-
-        q_tm1, _ = combine_dim(q_tm1)
-        q_t, _ = combine_dim(q_t)
-    else:
-        q_tm1 = q_values
-        q_tm1.cut_dimension(axis=1, end=-bootstrap_n)
-
-        q_t = target_q_values
-        q_t.cut_dimension(axis=1, start=bootstrap_n)
-    n_step_rewards = rewards[:, :-bootstrap_n]
-    n_step_discount = discounts[:, :-bootstrap_n]
-    for i in range(1, bootstrap_n + 1):
-        n_step_rewards += rewards[:, i : i + seq_len - bootstrap_n]
+    q_tm1 = q_values
+
+    # The last values that rolled over do not matter because a
+    # mask is applied to it (hopefully).
```

Comment:
I don't think it is good for us to have something happen "hopefully" 😅  Where is this expected to be handled? And by whom?

Reply:
This is handled already :) I will remove the hopefully.